src/NewRemoting/MessageHandler.cs

```diff
@@ -1266,14 +1289,60 @@ public static Exception DecodeException(BinaryReader reader, string otherSidePro
 			return root;
 		}
 
-		private static Exception DecodeExceptionInternal(BinaryReader reader, string otherSideProcessId)
+		private static (string Name, object Value) DecodeProperty(PropertyInfo property, BinaryReader reader, string otherSideProcessId, MessageHandler handler)
+		{
+			if (property != null)
+			{
+				object propertyVal;
+				if (property.PropertyType == typeof(string))
+				{
+					propertyVal = reader.ReadString();
+				}
+				else
+				{
+					try
+					{
+						propertyVal = handler?.ReadArgumentFromStream(reader, null, null, true,
+							property.PropertyType, otherSideProcessId, null);
+					}
+					catch (Exception)
+					{
+						propertyVal = null;
+					}
+				}
+
+				return (property.Name, propertyVal);
+			}
+
+			return (string.Empty, null);
+		}
+
+		private static Exception DecodeExceptionInternal(BinaryReader reader, string otherSideProcessId, MessageHandler handler)
 		{
 			string exceptionTypeName = reader.ReadString();
 			string remoteMessage = reader.ReadString();
 			int hresult = reader.ReadInt32();
 			string remoteStack = reader.ReadString();
 			Type exceptionType = Server.GetTypeFromAnyAssembly(exceptionTypeName);
 
+			var properties = exceptionType.GetProperties();
+			var propertyCount = properties.Count(x => x.CanWrite);

```

Comment:
I think it would be better to pass the number of properties along. That way, it should be guaranteed we read the same number of fields from the stream than what was written. This would otherwise fail seriously if there was a difference between the types on either side of the stream (e.g. a new field added).

Reply:
done